project-koku/koku
# COST-310: Pagination & Sorting fix for org_unit group_by

### Description
This is the second half of [COST-310](https://issues.redhat.com/browse/COST-310) (Fixing pagination & sorting when you are grouping by org_unit on the costs endpoint) 

The problem with pagination is that we cant limit through the queryset since we are adding multiple querysets together. The same problem prevails for ordering by cost. 

### Testing 
1. Run `make load-aws-org-unit-tree`

To test ordering by cost: 
Visit: 
* http://127.0.0.1:8000/api/cost-management/v1/reports/aws/costs/?group_by[org_unit_id]=R_001&order_by[cost]=asc
* http://127.0.0.1:8000/api/cost-management/v1/reports/aws/costs/?group_by[org_unit_id]=R_001&order_by[cost]=desc
For each day make sure the `org_entities` are in asc/desc order according to cost 
* http://127.0.0.1:8000/api/cost-management/v1/reports/aws/costs/?filter[resolution]=monthly&filter[time_scope_units]=month&filter[time_scope_value]=-1&group_by[org_unit_id]=R_001&order_by[cost]=asc
*  http://127.0.0.1:8000/api/cost-management/v1/reports/aws/costs/?filter[resolution]=monthly&filter[time_scope_units]=month&filter[time_scope_value]=-1&group_by[org_unit_id]=R_001&order_by[cost]=desc
For the month make sure the `org_entities` are in asc/desc order according to cost

To test pagination: 
* http://127.0.0.1:8000/api/cost-management/v1/reports/aws/costs/?filter[resolution]=monthly&filter[time_scope_units]=month&filter[time_scope_value]=-1&group_by[org_unit_id]=R_001&filter[limit]=1
* test increasing the offset to get the next results http://127.0.0.1:8000/api/cost-management/v1/reports/aws/costs/?filter[resolution]=monthly&filter[time_scope_units]=month&filter[time_scope_value]=-1&group_by[org_unit_id]=R_001&filter[limit]=1&filter[offset]=1
* test that increasing the filter gives you all the results http://127.0.0.1:8000/api/cost-management/v1/reports/aws/costs/?filter[resolution]=monthly&filter[time_scope_units]=month&filter[time_scope_value]=-1&group_by[org_unit_id]=R_001&filter[limit]=10
Check that the count is the number of org_entities & that the offset in the links is increased as expected

Also try this on the daily result level (make sure the count is the number of individual org units over the past 10 days & that the limit limits the number returned per day & that the links are correct for the next page) 
* http://127.0.0.1:8000/api/cost-management/v1/reports/aws/costs/?group_by[org_unit_id]=R_001&filter[limit]=1

To test the disabled wildcard (You should get a 400):
* [http://127.0.0.1:8000/api/cost-management/v1/reports/aws/costs/?group_by[org_unit_id]=*](http://127.0.0.1:8000/api/cost-management/v1/reports/aws/costs/?group_by[org_unit_id]=*)

### Smoke tests results
_============================= 1 failed, 3946 passed, 80 skipped, 1709 deselected in 1396.08s (0:23:16) =============================_
